gui/confirm.lua

```diff
@@ -1,8 +1,36 @@
 -- config ui for confirm
+--@ module = true
+local GLOBAL_KEY = 'confirm'
+local CONFIRM_CONFIG_FILE = 'dfhack-config/confirm.json'
+local json = require('json')
+confirm = require('plugins.confirm')
+gui = require('gui')
+widgets = require('gui.widgets')
 
-confirm = require 'plugins.confirm'
-gui = require 'gui'
-widgets = require 'gui.widgets'
+loaded = loaded or false
+
+if dfhack.onStateChange then
+    dfhack.onStateChange[GLOBAL_KEY] = function(_)
+        if loaded then
+            return
+        end
+
+        local ok, config = pcall(json.decode_file, CONFIRM_CONFIG_FILE)
+        local conf_data = ok and config or confirm.get_conf_data()
+        if ok then
+            for _, c in ipairs(conf_data) do
+                -- for some reason changing the conf state here doesn't seem to change the result
+                -- when the dialog is opened like it did when it was in the other script?
+                confirm.set_conf_state(c.id, c.enabled)
+            end
+        end
+        loaded = true
+    end
+end
```

Comment:
```suggestion
dfhack.onStateChange[GLOBAL_KEY] = function(sc)
    if sc ~= SC_CORE_INITIALIZED then return end

    local ok, config = pcall(json.decode_file, CONFIRM_CONFIG_FILE)
    if not ok then return end
    for _, c in ipairs(config) do
        confirm.set_conf_state(c.id, c.enabled)
    end
end
```

Reply:
what you had here should have worked, I think, I just cleaned it up a bit. Did you restart DF to test? it wouldn't have loaded until you caused a state change event.

Reply:
Yeah it wasn't working even with restarting DF. The file was properly written, and with print statements I could see that the `onStateChange` code was successfully being called. But then when I invoked `gui/confirm`, the result of `confirm.get_conf_data()` was not affected.

Reply:
Seeing the same thing with this version

 of the code. It totally seems like it should work but..

https://user-images.githubusercontent.com/2104555/218814231-0a5249e8-38dc-4abf-8785-4801a64b944e.mp4

Reply:
I experimented with it a bit. It seems like if the `onStateChange` code happens too early, the change to conf state doesn't persist.

If I change the `sc` code to `SC_MAP_LOADED`, the confirmations have the expected state when I load up a map and invoke `gui/confirm`. Similarly, if I remove the condition for `sc` code altogether, I can see the `onStateChange` code fire off twice as I load up the game, and then 3 more times once the game has finished loading (event happen in the order: SC_CORE_INITIALIZED, SC_VIEWSCREEN_CHANGED, SC_VIEWSCREEN_CHANGED, SC_VIEWSCREEN_CHANGED, SC_VIEWSCREEN_CHANGED). The confirmations then also have the expected state when I invoke `gui/confirm`.

Reply:
If I put:

```
loaded = loaded or false

if dfhack.onStateChange then
    dfhack.onStateChange[GLOBAL_KEY] = function(sc)
        if loaded then return end
        local ok, config = pcall(json.decode_file, CONFIRM_CONFIG_FILE)
        if not ok then return end
        for _, c in ipairs(config) do
            set_conf_state(c.id, c.enabled)
            print('Set ' .. c.id .. ' to ' .. tostring(c.enabled))
        end
        loaded = true
    end
end
```

Back into `plugins/confirm.lua`, everything works as expected. If I don't do the `if dfhack.onStateChange then` check, then I see this error in the console:

```
Loading script: hack/init/dfhack.tools.init
...p\DFHack Dev\Dwarf Fortress\hack\lua\plugins\confirm.lua:17: attempt to index a nil value (field 'onStateChange')
stack traceback:
        ...p\DFHack Dev\Dwarf Fortress\hack\lua\plugins\confirm.lua:17: in main chunk
        [C]: in function 'require'
```

But then it still works as expected:

![image](https://user-images.githubusercontent.com/2104555/218826399-9555defb-0391-483f-8952-d56981df6ced.png)

It definitely seems to be a matter of timing but I can't work out exactly what the right time is.

Reply:
let me look into why this isn't working as I expect it to. other scripts don't seem to have trouble attaching handlers when they are first loaded

Reply:
it looks like the confirm plugin itself is overwriting all of its state with "enabled" when it is enabled (which happens after core init). That part needs updating to work with the state restoration flow.

Reply:
Ahhh, okay yeah I think I see that now. Updating.

Reply:
I think I made the required update (to the other PR). I just need to make one more change to this PR now I think to ensure that confirms get enabled by default if config is not successfully loaded, since nothing else will should be enabling them now

Reply:
I think my latest commit handles this